agora/Agora/Governor.hs

```diff
@@ -1,62 +1,638 @@
+
+    ownOutputs <- plet $ findOutputsToAddress # txInfo' # selfAddress
+    passert "Exactly one utxo should be sent to the governor" $
+      plength # ownOutputs #== 1
+
+    ownOutput <- pletFields @'["value", "datumHash"] $ phead # ownOutputs
+    let ownOuputDatumNFTAmount = stateTokenValueOf # ownOutput.value
+    passert "State token should stay at governor's address" $
+      ownOuputDatumNFTAmount #== 1
+    passert "Output utxo to governor should have datum" $
+      pisDJust # ownOutput.datumHash
+
+    let outputGovernorStateDatumHash = mustBePDJust # "Output governor state datum hash not found" # ownOutput.datumHash
+
+    newDatumData <-
+      plet $
+        pforgetData $
+          pdata $
+            mustBePJust # "Ouput governor state datum not found"
+              #$ pfindDatum # outputGovernorStateDatumHash # txInfo'
+
+    case redeemer of
+      PCreateProposal _ -> P.do
+        let expectedNextProposalId = pnextProposalId # oldParams.nextProposalId
+            expectedNewDatum =
+              pcon $
+                PGovernorDatum $
+                  pdcons @"proposalThresholds" # oldParams.proposalThresholds
+                    #$ pdcons @"nextProposalId" # pdata expectedNextProposalId # pdnil
+
+        passert "Unexpected governor state datum" $
+          newDatumData #== pforgetData (pdata expectedNewDatum)
+
+        passert "Exactly one proposal token must be minted" $
+          hasOnlyOneTokenOfCurrencySymbol # pproposalSymbol # txInfo.mint
+
+        outputs <- plet $ findOutputsToAddress # ctx.txInfo # pproposalValidatorAddress
+        passert "Exactly one utxo should be sent to the proposal validator" $
+          plength # outputs #== 1
+
+        output <- pletFields @'["value", "datumHash"] $ phead # outputs
+        passert "The proposal state token must be sent to the proposal validator" $
+          psymbolValueOf # pproposalSymbol # output.value #== 1
+
+        passert "The utxo paid to the proposal validator must have datum" $
+          pisDJust # output.datumHash
+
+        outputProposalDatum' <-
+          plet $
+            mustFindDatum' @PProposalDatum
+              # output.datumHash
+              # ctx.txInfo
+
+        passert "Proposal datum must be valid" $
+          proposalDatumValid # outputProposalDatum'
+
+        proposalDatum <-
+          pletFields
+            @'["id", "status", "cosigners", "thresholds", "votes"]
+            outputProposalDatum'
+
+        passert "Invalid proposal id in proposal datum" $
+          proposalDatum.id #== oldParams.nextProposalId
+
+        passert "Invalid thresholds in proposal datum" $
+          proposalDatum.thresholds #== oldParams.proposalThresholds
+
+        passert "Initial proposal votes should be empty" $
+          pnull #$ pto $ pto $ pfromData proposalDatum.votes
+
+        -- TODO: should we check cosigners here?
+
+        let isProposalDraft = pmatch (pfromData proposalDatum.status) $ \case
```

Comment:
This could probably be as simple as `proposalDatum.status #== pconstantData Draft`, right?